### PR TITLE
Added support for running fake workers in the background

### PR DESCRIPTION
**Pull request by jonasfj**
Subset of https://github.com/dart-lang/pub-dev/pull/6169 -- planning to use this later.

### CODE REVIEW

File: app/lib/fake/server/fake_analyzer_service.dart [hunk 1]
```diff
@@ -57,8 +61,14 @@ class FakeAnalyzerService {
           });
           _logger.info('running on port $port');
 
+          (taskWorkerCloudCompute as FakeCloudCompute).startInstanceExecution();
+          await taskBackend.backfillTrackingState();
+          await taskBackend.start();
+
           await ProcessSignal.sigint.watch().first;
 
+          await taskBackend.stop();
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply app/lib/fake/server/fake_analyzer_service.dart:70]: Love it, that's a great solution :D

File: app/lib/task/cloudcompute/fakecloudcompute.dart [hunk 2]
```diff
@@ -104,6 +126,79 @@ class FakeCloudCompute extends CloudCompute {
...
+        _log.info('pub_worker exit code: $exitCode');
+      } finally {
+        // Don't terminate the instance if it's already deleted.
+        if (_instances.any((i) => i.instanceName == instance.instanceName)) {
+          fakeTerminateInstance(instance.instanceName);
+        }
+        _doneRunningInstance.complete();
+
+        // Start running the next instance
+        scheduleMicrotask(() async => await _keepExecutionLoopAlive());
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply app/lib/task/cloudcompute/fakecloudcompute.dart:176]: It's possible the invariant might be harder to follow.

In the same sync block we release `_doneRunningInstance.complete();` we'd to stop working, or overwrite it another completer...

And we'd end up repeating some of the if-statements we have the top...

In the current configuration it's always safe to call `_keepExecutionLoopAlive`. We only have to be careful when calling ourselves, to make sure that we've released the previous lock (`_doneRunningInstance`).

File: app/lib/fake/server/fake_analyzer_service.dart [hunk 3]
```diff
@@ -57,6 +62,11 @@ class FakeAnalyzerService {
           });
           _logger.info('running on port $port');
 
+          (taskWorkerCloudCompute as FakeCloudCompute).startInstanceExecution();
+          await taskBackend.backfillTrackingState();
+          await taskBackend.start();
+          registerScopeExitCallback(taskBackend.stop);
```
Comment (isoos) [inline app/lib/fake/server/fake_analyzer_service.dart:68]: Oh, I've just found out that `services.dart` already has it on line 263. Do we still need it here?

File: app/lib/task/cloudcompute/fakecloudcompute.dart [hunk 4]
```diff
@@ -104,6 +126,80 @@ class FakeCloudCompute extends CloudCompute {
...
+          mode: ProcessStartMode.inheritStdio,
+        );
+        final exitCode = await proc.exitCode;
+        _log.info('pub_worker exit code: $exitCode');
+      } finally {
+        // Don't terminate the instance if it's already deleted.
+        if (_instances.any((i) => i.instanceName == instance.instanceName)) {
+          fakeTerminateInstance(instance.instanceName);
+        }
+        // Must happen before we start the next instance.
```
Comment (isoos) [inline app/lib/task/cloudcompute/fakecloudcompute.dart:173]: Nit: can we do any assert here?

**Review by isoos — APPROVED** [review]

File: app/lib/task/cloudcompute/fakecloudcompute.dart (hunk 4 above)
Comment (jonasfj) [reply app/lib/task/cloudcompute/fakecloudcompute.dart:173]: I'm not sure how, we start the next iteration in the very next line. And the next iteration will refuse to run if it's not completed.

File: app/lib/fake/server/fake_analyzer_service.dart (hunk 3 above)
Comment (jonasfj) [reply app/lib/fake/server/fake_analyzer_service.dart]: good catch... there is no reason not call it -- also no harm in doing it. Remove it!

**Review by isoos — APPROVED** [review]

**Pull request merged**